### DISCUSSION TITLE
Welcome to WhatsApp-Key-Database-Extractor Discussions!

### DISCUSSION THREAD

**Discussion by YuvrajRaghuvanshiS** [YuvrajRaghuvanshiS/WhatsApp-Key-Database-Extractor#4] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Tkd-Alex** [comment]
Awesome projects! It's really funny, In the last 5 months I've worked on the same idea :rofl:

**Reply by YuvrajRaghuvanshiS** [reply]
Haha that's great! How much did you implement?

**Reply by Tkd-Alex** [reply]
I've a fork of the original project by EliteAndroidApps and add a reboot for prevent `[INSTALL_FAILED_VERSION_DOWNGRADE]`. That's it.

I've also worked on an interactive and 'remote" version of WhatsDump so you can extract the key remotely. It's my thesis for the master degree

**Reply by YuvrajRaghuvanshiS** [reply]
A reboot prevents [INSTALL_FAILED_VERSION_DOWNGRADE]? You just fixed [#16](https://github.com/YuvrajRaghuvanshiS/WhatsApp-Key-Database-Extractor/issues/16#issuecomment-768933461). Is there any explaination of why that so?

Remotely you mean over TCP? That a huge security risk. I tried implementing that in my project but adb did not allow backup over TCP

**Reply by Tkd-Alex** [reply]
Oh. I did not read your problem on #16 .
Yes for me a reboot works very well as you can see here:
https://github.com/Tkd-Alex/WhatsApp-Key-DB-Extractor/blob/50ecb9e7d5d9b20c8b63d3ce6c84b47d81e4d9b9/WhatsAppKeyDBExtract.sh#L105-L112

For WhatsDump we are working on the same way. I've a forked of your forked where I've ported the entire project to Py3 and then do some refactory. In another repo (currently private) I'm able to drive WhatsDump with HTTP Requests.

**Reply by YuvrajRaghuvanshiS** [reply]
@Tkd-Alex I have made it work with TCP and solved 549 bytes ab file issue (real pain in the a$$ tbh) you can check if you have same issue.

**Comment by YuvrajRaghuvanshiS** [comment]
> Yes for me a reboot works very well as you can see here:
https://github.com/Tkd-Alex/WhatsApp-Key-DB-Extractor/blob/50ecb9e7d5d9b20c8b63d3ce6c84b47d81e4d9b9/WhatsAppKeyDBExtract.sh#L105-L112

I see that I'm gonna implement that in issue branch of my project now with maybe add argument like `python wa_kdbe.py --allow-reboot`

> For WhatsDump we are working on the same way. I've a forked of your forked where I've ported the entire project to Py3 and then do some refactory. 

Well you beat me here LoL. I was making time to port that in python3 and fix some issues but couldn't because of university exams. Is you fork up and running? I want to add that in my README.md as a fallback method because on some devices my method just doesn't work for some reason.

> In another repo (currently private) I'm able to drive WhatsDump with HTTP Requests.

That's awesome where do you plan on hosting that?

**Reply by Tkd-Alex** [reply]
> Is you fork up and running?

As I know yes! It's currently working :)

> That's awesome where do you plan on hosting that?

For the moment idk. But I think I'll release on public on the May 2021.

**Reply by YuvrajRaghuvanshiS** [reply]
> As I know yes! It's currently working :)

Can I link that in my project?

**Reply by YuvrajRaghuvanshiS** [reply]
> For the moment idk. But I think I'll release on public on the May 2021.

Looking forward to it.

**Reply by Tkd-Alex** [reply]
> > As I know yes! It's currently working :)
> 
> Can I link that in my project?

You are welcome!

**Comment by jinucompute** [comment]
When try this i get the error 

failed to install tmp/LegacyWhatsApp.apk: Failure [-26: Package com.whatsapp new target SDK 19 doesn't support runtime permissions but the old target SDK 23 does.]

Any ideas how to solve this

**Reply by YuvrajRaghuvanshiS** [reply]
Create an issue for this.

**Comment by LucD20** [comment]
I'm very new at this so I'm not sure how to do much. I'm on Windows 10, downloaded everything needed and tried to open wa_kdbe.py, but I get "'pip3' is not recognized as an internal or external command, operable program or batch file" 5 times before the window closes. Any idea what I'm doing wrong?

**Reply by YuvrajRaghuvanshiS** [reply]
run
```
pip install --upgrade termcolor wget packaging
```
I have made it to install these automatically but your system is not supporting that. Also check which python you have, it works with python3 only.

**Reply by LucD20** [reply]
Thank you for taking your time to answer! I have the newest python (3.9.2) installed. How do I run these commands? Like, on CMD, do I have to open Python etc

**Reply by YuvrajRaghuvanshiS** [reply]
Clone the repo. Go inside "Whatsapp-Key-Database-Extractor" folder. Open cmd there and run "python wa_kdbe.py", make sure to connect phone via usb cable and usb debugging is on in android developer options.
If you face any issues during process like `[INSTALL_FAILED_VERSION_DOWNGRADE]` run with allow-reboot flag `python wa_kdbe.py --allow-reboot`

**Reply by YuvrajRaghuvanshiS** [reply]
Also that pip command can run anywhere those are just python modules/libraries. Make sure to run that before run main script.

**Reply by LucD20** [reply]
Thanks again for the help.
For some reason, when I run "python wa_kdbe.py" I get an error, saying: "python is not recognized as an internal or external command operable program or batch file". Do you know what could be causing this?

**Reply by YuvrajRaghuvanshiS** [reply]
Yes, python is not in PATH. Try running with `python3 wa_kdbe.py` If that is also not working you need to reinstall python3 and during installation tick the checkbox saying "Add python to PATH"

**Comment by lqpl-l** [comment]
Hi guys,
I have this on cmd screen. Any advice?


Installing legacy WhatsApp 2.11.431
61 KB/s (251 bytes in 0.003s)
Error: Failed to parse APK file: /data/local/tmp/LegacyWhatsApp.apk

Exception occurred while executing:
java.lang.IllegalArgumentException: Error: Failed to parse APK file: /data/local/tmp/LegacyWhatsApp.apk
at com.android.server.pm.PackageManagerShellCommand.setParamsSize(PackageManagerShellCommand.java:338)
at com.android.server.pm.PackageManagerShellCommand.runInstall(PackageManagerShellCommand.java:906)
at com.android.server.pm.PackageManagerShellCommand.onCommand(PackageManagerShellCommand.java:158)
at android.os.ShellCommand.exec(ShellCommand.java:103)
at com.android.server.pm.PackageManagerService.onShellCommand(PackageManagerService.java:21826)
at android.os.Binder.shellCommand(Binder.java:634)
at android.os.Binder.onTransact(Binder.java:532)
at android.content.pm.IPackageManager$Stub.onTransact(IPackageManager.java:2796)
at com.android.server.pm.PackageManagerService.onTransact(PackageManagerService.java:3985)
at android.os.Binder.execTransact(Binder.java:731)
Caused by: android.content.pm.PackageParser$PackageParserException: Failed to parse /data/local/tmp/LegacyWhatsApp.apk
at android.content.pm.PackageParser.parseApkLiteInner(PackageParser.java:1584)
at android.content.pm.PackageParser.parseApkLite(PackageParser.java:1569)
at com.android.server.pm.PackageManagerShellCommand.setParamsSize(PackageManagerShellCommand.java:331)
... 9 more
Install complete
Press any key to continue . . .


Thank you!

**Reply by YuvrajRaghuvanshiS** [reply]
TCP or USB mode?

**Reply by YuvrajRaghuvanshiS** [reply]
If "[INSTALL_PARSE_FAILED_NOT_APK]" delete helpers/LegacyWhatsApp.apk and re-run.

**Reply by lqpl-l** [reply]
USB mode

**Comment by lqpl-l** [comment]
Thank you for your help mate! Really appreciate!
Still the same, I have Android 9, so it asks to reboot, I reboot but still the same... Do you need log file?

**Comment by lqpl-l** [comment]
Does it matter where the program folder is? Any special location on the disk etc?

**Reply by YuvrajRaghuvanshiS** [reply]
nah

**Comment by lqpl-l** [comment]
This is the log:
##### WhatsApp installation and version checks #####
WhatsApp 2.21.6.17 installed
WhatsApp downgrade required
Found legacy WhatsApp 2.11.431 in local folder

Backing up WhatsApp 2.21.6.17
/data/app/com.whatsapp-FDf41Xm7AmxuFbQpbGaw-w==/base.apk: 1 file pulled. 31.3 MB/s (32365604 bytes in 0.987s)
Backup complete

Removing WhatsApp 2.21.6.17 skipping data
Success
Removal complete

Installing legacy WhatsApp 2.11.431
Android version 7.0 or higher detected
Device may have to be rebooted to downgrade and avoid failure
like "[INSTALL_FAILED_VERSION_DOWNGRADE]"

Do you want to reboot (Y/N)? y

Press any key once device is ready (to avoid error "can't find package ...")
Press any key to continue . . .
Performing Streamed Install
adb: failed to install tmp\LegacyWhatsApp.apk: Failure [INSTALL_PARSE_FAILED_NOT_APK: Failed to parse /data/app/vmdl642722234.tmp/base.apk]

Exiting ...

**Reply by YuvrajRaghuvanshiS** [reply]
This is not my repo's log. Clone this one.

**Reply by xMoAmine** [reply]
Can you perhaps answer this question so I can faintly understand what they back up https://github.com/EliteAndroidApps/WhatsApp-Key-DB-Extractor/issues/41#issue-992764625 like is it in any way detrimental to the device itself? Like where does this back up go

**Reply by xMoAmine** [reply]
@YuvrajRaghuvanshiS

**Reply by YuvrajRaghuvanshiS** [reply]
It backs up app data including it's private data (/data/data/<package_name>), shared data etc. It does not get stored on device you need another device to run adb commands there only it gets stored like your computer (although you can rig termux so that it can save that .ab file in device itself.)
Talking of security android has already announced that android backup is deprecated in version 10 (I believe) and other than that apps now a days include a line in their manifest file (allowBackup=False) which saves them from being backed up even on older versions of android.

Full backup implies full backup of that app not entire phone, once you tap backup there will a line below password text area showing you the name of app being backed up (whether it is backing up one app or entire phone, in case of complete backup it will show you apps one by one whichever it is backing up). As long as you don't connect you phone to strangers' laptop and keep USB debugging off all the time you are safe. Also there is an option in adb (adb via tcp) which allows one to run adb commands remotely even if you are on different networks. So my advice is keep USB debugging off in developers options, reboot your phone once a day (doing so assigns you a new IP), and tap revoke USB debugging authorisation.

I might have got out of context here but that was all crucial information.

**Reply by xMoAmine** [reply]
Man you're a legend @YuvrajRaghuvanshiS Thanks for explaining.

One last question if you don't mind! What happens if you backup the app for a phone with android 10 or higher without downloading the ADB drivers, like just running the DB extractor .bat file. Does it only download the legacy whatsapp apk and do nothing or?

**Reply by YuvrajRaghuvanshiS** [reply]
adb is a bridge. Without it you would not be able to communicate with android. 
> Does it only download the legacy whatsapp apk and do nothing or?

What you're asking here is regarding someone else's work. I need to go through their source code to find out in what order they are doing stuff. I ship my project with adb built in.

Also as far as my knowledge on this subject goes there should be either one of two things necessary for backup to work.
1. Android is lower version which supports backup
2. Application itself allows to be backed up.
 
I have tested my tool without Android 11 also with no fail whatsoever.

**Reply by xMoAmine** [reply]
Ahh so I tried out running the DB extractor which installs a temporary legacy whatsapp APK to decrypt a crypt file, but after saying it was done and removed the Legacy Whatsapp APK, nothing came into the extracted folder, this can only mean the backup wasn't a succes right?

**Reply by YuvrajRaghuvanshiS** [reply]
Again you're referencing TripCode's code I believe. You can clone this one and ask me anything but I am really not a person when it comes to answer his codes' questions.

**Reply by xMoAmine** [reply]
Ahh no I was referencing the thread where you posted the link to this repo, the official uhh Whatsapp Key?DB extractor, this one: https://forum.xda-developers.com/t/tool-whatsapp-key-db-extractor-crypt6-12-non-root-updated-october-2016.2770982/page-34

**Reply by YuvrajRaghuvanshiS** [reply]
This was similar project so I posted. But mine in entirely different in terms of language and capabilities. The bat file is related to TripCode's repo not mine.

**Reply by createdroid** [reply]
> Este fue un proyecto similar, así que publiqué. Pero el mío es completamente diferente en términos de lenguaje y capacidades. El archivo bat está relacionado con el repositorio de TripCode, no con el mío.

Hello, but tell us, could you finish your project or still nothing?

**Comment by yasser2016** [comment]
still hanging on hit enter to continue for more than 10min / Android 12
[New Microsoft Word Document.docx](https://github.com/YuvrajRaghuvanshiS/WhatsApp-Key-Database-Extractor/files/8309305/New.Microsoft.Word.Document.docx)


![image](https://user-images.githubusercontent.com/23379915/159123028-752b1e93-c3b9-47c1-a939-e139da3e12a2.png)

**Comment by tiago-204** [comment]
HI! Great proyect, but i have a problem 

[WinError 2] El sistema no puede encontrar el archivo especificado: 'tmp/whatsapp.tar'

IDK why... any suggestion?
